File: blaise-common/src/main/java/com/googlecode/blaisemath/util/AnchoredImage.java
/**
 * AnchoredImage.java
 * Created on Sep 12, 2014
 */
package com.googlecode.blaisemath.util;

/*
 * #%L
 * blaise-common
 * --
 * Copyright (C) 2014 - 2017 Elisha Peterson
 * --
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */

import java.awt.Image;
import java.awt.geom.Rectangle2D;
import java.awt.image.ImageObserver;
import javax.annotation.Nullable;

/**
 * An image anchored at a given location.
 * 
 * @author petereb1
 */
public final class AnchoredImage extends Point2DBean {

    private final Image originalImage;
    private final Image scaledImage;
    private final String ref;
    private final java.lang.Double width;
    private final java.lang.Double height;

    public AnchoredImage(double x, double y, Image image, @Nullable String ref) {
        this(x, y, null, null, image, ref);
    }

    public AnchoredImage(double x, double y, @Nullable java.lang.Double width, @Nullable java.lang.Double height, Image image, @Nullable String ref) {
        super(x, y);
        this.width = width;
        this.height = height;
        this.ref = ref;
        this.originalImage = image;
        if (width != null && width > 0 && height != null && height > 0 && (image.getWidth(null) != width || image.getHeight(null) != height)) {
            this.scaledImage = image.getScaledInstance(width.intValue(), height.intValue(), Image.SCALE_DEFAULT);
        } else {
            this.scaledImage = image;
        }
    }

    //<editor-fold defaultstate="collapsed" desc="PROPERTIES">
    //
    // PROPERTIES
    //

    public String getReference() {
        return ref;
    }

    public double getWidth() {
        return width == null ? scaledImage.getWidth(null) : width;
    }

    public double getHeight() {
        return height == null ? scaledImage.getHeight(null) : height;
    }
    
    public Rectangle2D getBounds(ImageObserver io) {
        double iw = width == null ? scaledImage.getWidth(io) : width;
        double ih = height == null ? scaledImage.getHeight(io) : height;
        return new Rectangle2D.Double(x, y, iw, ih);
    }
    
    public Image getImage() {
        return scaledImage;
    }
    
    public Image getOriginalImage() {
        return originalImage;
    }
    
    //</editor-fold>
    
}

File: blaise-graphics-kt/src/main/java/com/googlecode/blaisemath/graphics/AnchoredImage.kt
package com.googlecode.blaisemath.graphics

import com.googlecode.blaisemath.coordinate.HasPoint2D
import com.googlecode.blaisemath.util.geom.Rectangle2
import java.awt.Image
import java.awt.image.ImageObserver

/*
* #%L
* BlaiseGraphics
* --
* Copyright (C) 2009 - 2021 Elisha Peterson
* --
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*      http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
* #L%
*/

/** An image anchored at a given location. */
class AnchoredImage(x: kotlin.Double, y: kotlin.Double, private val _width: kotlin.Double? = null, private val _height: kotlin.Double? = null,
                    val originalImage: Image, val reference: String?) : HasPoint2D(x, y) {

    val width: kotlin.Double
        get() = _width ?: image.getWidth(null).toDouble()
    val height: kotlin.Double
        get() = _height ?: image.getHeight(null).toDouble()

    val image: Image = image()

    private fun image(): Image {
        val widthPos = _width != null && _width > 0
        val heightPos = _height != null && _height > 0
        return if (widthPos && heightPos && (originalImage.getWidth(null).toDouble() != _width || originalImage.getHeight(null).toDouble() != _height)) {
            originalImage.getScaledInstance(_width!!.toInt(), _height!!.toInt(), Image.SCALE_DEFAULT)
        } else {
            originalImage
        }
    }

    fun getBounds(io: ImageObserver?) = Rectangle2(x, y, _width ?: image.getWidth(io).toDouble(), _height ?: image.getHeight(io).toDouble())

}